pkg/events/event_streaming.go

```diff
@@ -0,0 +1,192 @@
+/*
+ Licensed to the Apache Software Foundation (ASF) under one
+ or more contributor license agreements.  See the NOTICE file
+ distributed with this work for additional information
+ regarding copyright ownership.  The ASF licenses this file
+ to you under the Apache License, Version 2.0 (the
+ "License"); you may not use this file except in compliance
+ with the License.  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package events
+
+import (
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/apache/yunikorn-core/pkg/log"
+	"github.com/apache/yunikorn-scheduler-interface/lib/go/si"
+)
+
+const defaultChannelBufSize = 100
+
+type EventStreaming struct {
+	buffer       *EventRingBuffer
+	stopCh       chan struct{}
+	eventStreams map[EventStream]consumerDetails
+	sync.Mutex
+}
+
+type consumerDetails struct {
+	local     chan *si.EventRecord
+	consumer  chan<- si.EventRecord
+	stopCh    chan struct{}
+	name      string
+	createdAt time.Time
+}
+
+// EventStream the type returned to the client who wants to capture the stream of events.
+type EventStream struct {
+	events <-chan si.EventRecord
+}
+
+type EventHistoryOpts struct {
+	interval   time.Duration
+	eventCount int
+}
+
+// PublishEvent publishes an event to event stream consumers. Events are first stored in a ring buffer,
+// then all clients are notified. Clients whose buffer is full are considered slow and removed.
+func (e *EventStreaming) PublishEvent(event *si.EventRecord) {
+	e.Lock()
+	defer e.Unlock()
+
+	e.buffer.Add(event)
+	for consumer, details := range e.eventStreams {
```

Comment:
We can add another level of channel communication here, so updating clients will occur on a separate goroutine. So this method can look like:

```
func (e *EventStreaming) PublishEvent(event *si.EventRecord) {
	e.Lock()
	defer e.Unlock()
	e.buffer.Add(event)
        e.events <- event
}
```

So the event flow will be:
incoming event --> `e.events` --> `consumerDetails.local` --> `consumerDetails.consumer`.

It's a matter of taste, this method is already fast and not affected by slow clients.